File: src/main/java/org/noxet/noxetserver/commands/social/Unblock.java
package org.noxet.noxetserver.commands.social;

import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import org.noxet.noxetserver.UsernameStorageManager;
import org.noxet.noxetserver.messaging.ErrorMessage;
import org.noxet.noxetserver.messaging.SuccessMessage;
import org.noxet.noxetserver.playerdata.PlayerDataManager;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class Unblock implements TabExecutor {
    @Override
    @SuppressWarnings("NullableProblems")
    public boolean onCommand(CommandSender commandSender, Command command, String s, String[] strings) {
        if(!(commandSender instanceof Player)) {
            new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Only players can unblock players.").send(commandSender);
            return true;
        }

        Player player = (Player) commandSender;

        if(strings.length == 0) {
            new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Missing argument: player to unblock.").send(player);
            return true;
        }

        UUID uuidToUnblock = new UsernameStorageManager().getUUIDFromUsernameOrUUID(strings[0]);

        if(uuidToUnblock == null) {
            new ErrorMessage(ErrorMessage.ErrorType.COMMON, "That player is not registered.").send(player);
            return true;
        }

        String unblockName = UsernameStorageManager.getCasedUsernameFromUUID(uuidToUnblock); // This works anyway.

        PlayerDataManager playerDataManager = new PlayerDataManager(player);

        if(!playerDataManager.doesContain(PlayerDataManager.Attribute.BLOCKED_PLAYERS, uuidToUnblock.toString())) {
            new ErrorMessage(ErrorMessage.ErrorType.COMMON, "You have not blocked " + unblockName + ".").send(player);
            return true;
        }

        playerDataManager.removeFromStringList(PlayerDataManager.Attribute.BLOCKED_PLAYERS, uuidToUnblock.toString()).save();

        new SuccessMessage(unblockName + " is no longer blocked. They player can now message, friend request and TPA you.").send(player);

        return true;
    }

    @Override
    @SuppressWarnings("NullableProblems")
    public List<String> onTabComplete(CommandSender commandSender, Command command, String s, String[] strings) {
        if(!(commandSender instanceof Player))
            return null;

        Player player = (Player) commandSender;

        List<String> completions = new ArrayList<>();

        if(strings.length == 1) {
            //noinspection unchecked
            for(String blockedUUIDString : (List<String>) new PlayerDataManager(player.getUniqueId()).get(PlayerDataManager.Attribute.BLOCKED_PLAYERS)) {
                UUID blockedUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(blockedUUIDString);
                completions.add(blockedUUID != null ? UsernameStorageManager.getCasedUsernameFromUUID(blockedUUID) : blockedUUIDString);
            }
        }

        return completions;
    }
}

File: src/main/java/org/noxet/noxetserver/commands/social/Unblock.scala
package org.noxet.noxetserver.commands.social

import org.bukkit.command.{Command, CommandSender, TabExecutor}
import org.bukkit.entity.Player
import org.noxet.noxetserver.commands.{PlayerTabExecutor, RegisteredCommand}
import org.noxet.noxetserver.messaging.ErrorMessage.ErrorType.*
import org.noxet.noxetserver.messaging.{ErrorMessage, SuccessMessage}
import org.noxet.noxetserver.playerdata.PlayerDataManager
import org.noxet.noxetserver.util.{PlayerIntel, UsernameStorageManager}

import java.util.UUID

object Unblock extends PlayerTabExecutor, RegisteredCommand("unblock", this):
  override def onPlayerCommand(player: Player, command: Command, s: String, strings: Array[String]): Boolean =
    strings match
      case Array(playerIdentifier, _*) =>
        PlayerIntel(playerIdentifier) match
          case Some(playerToUnblock) =>
            val playerDataManager = PlayerDataManager(player)

            if !playerDataManager.doesContain(PlayerDataManager.Attribute.BlockedPlayers, playerToUnblock.uuid.toString) then
              ErrorMessage(Common, s"You have not blocked $playerToUnblock.").send(player)
              return true

            playerDataManager.removeFromStringList(PlayerDataManager.Attribute.BlockedPlayers, playerToUnblock.uuid.toString).save()

            SuccessMessage(s"$playerToUnblock is no longer blocked. They can now message, friend request and TPA you.").send(player)

            true
          case None =>
            ErrorMessage(Common, "That player is not registered.").send(player)
            true
      case _ =>
        ErrorMessage(Argument, "Missing argument: player to unblock.").send(player)
        true

  override def onPlayerTabComplete(player: Player, command: Command, s: String, strings: Array[String]): List[String] =
    strings match
      case Array(_) =>
        PlayerDataManager(player)
          .get(PlayerDataManager.Attribute.BlockedPlayers).asInstanceOf[List[String]]
          .flatMap(PlayerIntel)
          .map(_.username)
      case _ => List()